From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Patrick Hemmer <patrick.hemmer@gmail.com>
Date: Tue, 11 Aug 2020 22:11:53 +0300
Subject: [PATCH] Use block distance in portal search radius

Original author: Patrick Hemmer <patrick.hemmer@gmail.com>

Previously when searching for a portal, the server would use the search radius to determine which chunks it needs to search, and then would
consider any portal within those chunks as a match, even if it is outside the search radius.
This changes it so that the portal list is filtered to only include portals within the configured search radius (ignoring the Y coordinate).

Ported to Yatopia and added per world config option by MrIvanPlays <ivan@mrivanplays.com>

diff --git a/src/main/java/de/minebench/origami/OrigamiConfig.java b/src/main/java/de/minebench/origami/OrigamiConfig.java
index 58bc509119847f42d96470285c9f5ffd69246771..5b834624f1bdd2d4096179d48d09e813e91de5ae 100644
--- a/src/main/java/de/minebench/origami/OrigamiConfig.java
+++ b/src/main/java/de/minebench/origami/OrigamiConfig.java
@@ -160,6 +160,11 @@ public final class OrigamiConfig {
                             && pillagerCollisions && ironGolemCollisions && miscCollisions && itemCollisions
                             && waterCreatureCollisions && waterAmbientCollisions;
         }
+
+        public boolean useBlockDistanceInPortalSearchRadius = false;
+        private void useBlockDistanceInPortalSearchRadius() {
+            useBlockDistanceInPortalSearchRadius = getBoolean("use-block-distance-in-portal-search-radius", false);
+        }
         // Yatopia end
     }
 
diff --git a/src/main/java/net/minecraft/server/PortalTravelAgent.java b/src/main/java/net/minecraft/server/PortalTravelAgent.java
index 67018d2abcca8f9dd71fc3a6bfff19c85e8bcef0..d861368550a8811eb442c30c508023f6029daa4a 100644
--- a/src/main/java/net/minecraft/server/PortalTravelAgent.java
+++ b/src/main/java/net/minecraft/server/PortalTravelAgent.java
@@ -29,6 +29,15 @@ public class PortalTravelAgent {
             return villageplacerecord.f().getY();
         });
         java.util.List<VillagePlaceRecord> list = villageplace.b(type -> type == VillagePlaceType.v, blockposition, i, VillagePlace.Occupancy.ANY).collect(java.util.stream.Collectors.toList());
+        // Yatopia start - fix portal bug
+        if (world.origamiConfig.useBlockDistanceInPortalSearchRadius) {
+            list.removeIf(villagePlaceRecord -> {
+                BlockPosition portalPosition = villagePlaceRecord.getPosition();
+                return Math.abs(portalPosition.getX() - blockposition.getX()) > i
+                        || Math.abs(portalPosition.getZ() - blockposition.getZ()) > i;
+            });
+        }
+        // Yatopia end
         Optional<VillagePlaceRecord> optional = Optional.empty();
         if (!list.isEmpty()) {
             list.sort(comparator);
